src/frameworks/compose/discover/frameworkMatcher.ts

```diff
@@ -0,0 +1,101 @@
+import { FirebaseError } from "../../../error";
+import { FrameworkSpec, FileSystem } from "./types";
+import { logger } from "../../../logger";
+
+export function filterFrameworksWithDependencies(
+  allFrameworkSpecs: FrameworkSpec[],
+  dependencies: Record<string, string>
+): FrameworkSpec[] {
+  return allFrameworkSpecs.filter((framework) => {
+    return framework.requiredDependencies.every((dependency) => {
+      return dependency.name in dependencies;
+    });
+  });
+}
+
+export async function filterFrameworksWithFiles(
+  allFrameworkSpecs: FrameworkSpec[],
+  fs: FileSystem
+): Promise<FrameworkSpec[]> {
+  try {
+    const filteredFrameworks = [];
+    for (const framework of allFrameworkSpecs) {
+      if (!framework.requiredFiles) {
+        filteredFrameworks.push(framework);
+        continue;
+      }
+      let isRequired = true;
+      for (let files of framework.requiredFiles) {
+        files = Array.isArray(files) ? files : [files];
+        for (const file of files) {
+          isRequired = isRequired && (await fs.exists(file));
+          if (!isRequired) {
+            break;
+          }
+        }
+      }
+      if (isRequired) {
+        filteredFrameworks.push(framework);
+      }
+    }
+
+    return filteredFrameworks;
+  } catch (error) {
+    logger.error("Error: Unable to filter frameworks based on required files", error);
+    throw error;
+  }
+}
+
+/**
+ * Embeded frameworks help to resolve tiebreakers when multiple frameworks are discovered.
+ * Ex: "next" embeds "react", so if both frameworks are discovered,
+ * we can suggest "next" commands by removing embeded framework(react).
+ */
+export function removeEmbededFrameworks(allFrameworkSpecs: FrameworkSpec[]): FrameworkSpec[] {
+  const embededFrameworkSet: Set<string> = new Set<string>();
+
+  allFrameworkSpecs.forEach((framework) => {
```

Comment:
Here and below: prefer `for (.. of ..)` instead of `.forEach()`

go/tsjs-style#iterating-containers